### PR TITLE
[DPE-5711] Add warnings to destructive actions

### PR DESCRIPTION
**Pull request by sinclert-canonical**
This PR adds warning level messages when a _destructive_ action is performed. The definition of _destructive_ seems to be associated with data loss, which I think it can only happen, from user action, whenever they force an upgrade when the charm has already detected its lack of compatibility.

### Additional info
- Paulo confirmed that this ticket was created to familiarised myself with the codebases.
- Alex referenced this PostgreSQL operator [PR](https://github.com/canonical/postgresql-k8s-operator/pull/753) as the reason behind the need for warnings.

### CODE REVIEW

**Review by taurus-forever — APPROVED** [review]

**Review by taurus-forever — COMMENTED** [review]
BTW, I would also warn:
```
15:17:20 ✔ taurus:(main)~/canonical/mysql-router-operator$ rg rmtree
src/container.py
50:    def rmtree(self):

src/snap.py
153:    def rmtree(self):
154:        shutil.rmtree(self)

lib/charms/tempo_coordinator_k8s/v0/charm_tracing.py
220:                shutil.rmtree(path)

src/workload.py
151:        self._container.router_config_directory.rmtree()
153:        self._router_data_directory.rmtree()

tests/unit/conftest.py
111:    monkeypatch.setattr("snap._Path.rmtree", lambda *args, **kwargs: None)
15:17:21 ✔ taurus:(main)~/canonical/mysql-router-operator$ 
```

**Comment by sinclert-canonical** [comment]
My interpretation was that I should only warn about _potential data loss / some destruction_ (from the [JIRA ticket](https://warthogs.atlassian.net/browse/DPE-5711)). To the best of my understanding, the [`Workload._disable_router`](https://github.com/canonical/mysql-router-operator/blob/0f42207bd911a925dc969394bfd7bc479b33dd8f/src/workload.py#L146-L155) method only removes access, but it does not produce data-loss. This method is even called during "normal" actions such as `upgrade` ([reference](https://github.com/canonical/mysql-router-operator/blob/0f42207bd911a925dc969394bfd7bc479b33dd8f/src/workload.py#L431-L432)).

I am open to also warn about those cases, if the criteria gets extended to cover both _data-loss_ and _access-loss_ scenarios. Do you want me to?

**Comment by paulomach** [comment]
> My interpretation was that I should only warn about _potential data loss / some destruction_ (from the [JIRA ticket](https://warthogs.atlassian.net/browse/DPE-5711)). To the best of my understanding, the [`Workload._disable_router`](https://github.com/canonical/mysql-router-operator/blob/0f42207bd911a925dc969394bfd7bc479b33dd8f/src/workload.py#L146-L155) method only removes access, but it does not produce data-loss. This method is even called during "normal" actions such as `upgrade` ([reference](https://github.com/canonical/mysql-router-operator/blob/0f42207bd911a925dc969394bfd7bc479b33dd8f/src/workload.py#L431-L432)).
> 
> I am open to also warn about those cases, if the criteria gets extended to cover both _data-loss_ and _access-loss_ scenarios. Do you want me to?

Besides test and lib code, Alex do have a point. Even though the method is called on "normal" handling, we cannot assume it will not be called unexpectedly due some bug, and that's exactly when we most want to have it logged.

**Review by paulomach — APPROVED** [review]

**Comment by sinclert-canonical** [comment]
Added info messages to [Workload._diable_router](https://github.com/canonical/mysql-router-operator/blob/0f42207bd911a925dc969394bfd7bc479b33dd8f/src/workload.py#L146-L155) method.

File: src/workload.py [hunk 1]
```diff
@@ -145,14 +145,14 @@ def _disable_tls(self) -> None:
 
     def _disable_router(self) -> None:
         """Disable router and clean up corresponding router files."""
-        logger.debug("Disabling MySQL Router service")
+        logger.info("Disabling MySQL Router service")
```
[the remark this answers is not in the document]
Comment (sinclert-canonical) [reply src/workload.py:148]: Hey Carl. I noticed you left the same concerns on this open PR [comment](https://github.com/canonical/mysql-router-k8s-operator/pull/336#discussion_r1869063784).

Let's continue the conversation there.

**Pull request merged**